Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.


In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

In [3]:
df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

#### Identify the dimensions of the dataset by determining the number of rows and columns it contains.

In [4]:
num_rows, num_columns = df.shape

# Print the dimensions
print("Number of rows:", num_rows)
print("Number of columns:", num_columns)

Number of rows: 4008
Number of columns: 11


#### Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. 

In [5]:
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

#### You should also provide suggestions for fixing any incorrect data types.

Customer Lifetime Value should be a float type.

In [6]:
df['Customer Lifetime Value'] = df['Customer Lifetime Value'].str.replace('%', '')

df['Customer Lifetime Value'] = df['Customer Lifetime Value'].astype(float)

#### Identify the number of unique values for each column

In [7]:
df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

####  determine which columns appear to be categorical

In [8]:
data_types = df.dtypes

# Determine which columns appear to be categorical
categorical = data_types[data_types == 'object'].index.tolist()
categorical

['Customer',
 'ST',
 'GENDER',
 'Education',
 'Number of Open Complaints',
 'Policy Type',
 'Vehicle Class']

In [9]:
df.isna().sum()


Customer                     2937
ST                           2937
GENDER                       3054
Education                    2937
Customer Lifetime Value      2940
Income                       2937
Monthly Premium Auto         2937
Number of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64

In [16]:
df.columns = df.columns.str.lower()


In [20]:
df.columns.tolist()

['customer',
 'st',
 'gender',
 'education',
 'customer lifetime value',
 'income',
 'monthly premium auto',
 'number of open complaints',
 'policy type',
 'vehicle class',
 'total claim amount']

In [17]:
categorical = df.select_dtypes(include=['object'])

numerical = df.select_dtypes(include=['int', 'float'])

#### You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.

In [18]:
for column in categorical.columns:
    unique_values = categorical[column].unique()
    print(f"Column '{column}': {unique_values}")

Column 'customer': ['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan]
Column 'st': ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
Column 'gender': [nan 'F' 'M' 'Femal' 'Male' 'female']
Column 'education': ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
Column 'number of open complaints': ['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan]
Column 'policy type': ['Personal Auto' 'Corporate Auto' 'Special Auto' nan]
Column 'vehicle class': ['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]


In [19]:
numerical_columns = df.select_dtypes(include=[int, float]).columns
range_of_values = df[numerical_columns].describe().loc[['min', 'max']]

# Print the range of values for numerical columns
print("Range of values for numerical columns:")
print(range_of_values)

Range of values for numerical columns:
     customer lifetime value   income  monthly premium auto  \
min                200435.07      0.0                  61.0   
max               5816655.35  99960.0               35354.0   

     total claim amount  
min            0.382107  
max         2893.239678  


 Insights: There seems to be values that have the same meaning in the same column.

#### Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. 

In [13]:
# Compute summary statistics for numerical columns
summary_statistics = df.describe()

# Calculate the mode for each numerical column
mode_values = df.mode().iloc[0]

# Add mode values to the summary statistics DataFrame
summary_statistics.loc['mode'] = mode_values

# Print the summary statistics
print("Summary Statistics for Numerical Columns:")
print(summary_statistics)


Summary Statistics for Numerical Columns:
       Customer Lifetime Value        Income  Monthly Premium Auto  \
count             1.068000e+03   1071.000000           1071.000000   
mean              7.936903e+05  39295.701214            193.234360   
std               6.434784e+05  30469.427060           1601.190369   
min               2.004351e+05      0.000000             61.000000   
25%               4.034080e+05  14072.000000             68.000000   
50%               5.881742e+05  36234.000000             83.000000   
75%               8.962872e+05  64631.000000            109.500000   
max               5.816655e+06  99960.000000          35354.000000   
mode              2.514592e+05      0.000000             65.000000   

       Total Claim Amount  
count         1071.000000  
mean           404.986909  
std            293.027260  
min              0.382107  
25%            202.157702  
50%            354.729129  
75%            532.800000  
max           2893.239678  
mode 

#### Provide your conclusions based on these summary statistics.

The dataset contains information about customers, including their Customer Lifetime Value, Income, Monthly Premium Auto, and Total Claim Amount. The average CLV is high, with significant variability. Income has moderate average but some zero values raise concerns. Monthly premiums show wide variation, and there are low-value claims alongside large ones. 

#### Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [14]:
# Compute for summary statistics for categorical columns
summary_categorical = {}

for column in categorical.columns:
    column_value_counts = categorical[column].value_counts()
    summary_categorical[column] = {
        'unique_values': len(column_value_counts),
        'most_frequent_value': column_value_counts.index[0],
        'count_most_frequent': column_value_counts.iloc[0],
    }

# summary statistics for categorical columns
print("Summary Statistics for Categorical Columns:")
for column, stats in summary_categorical.items():
    print(f"Column '{column}':")
    print(f"  Unique values: {stats['unique_values']}")
    print(f"  Most frequent value: {stats['most_frequent_value']}")
    print(f"  Count of most frequent value: {stats['count_most_frequent']}")


Summary Statistics for Categorical Columns:
Column 'Customer':
  Unique values: 1071
  Most frequent value: RB50392
  Count of most frequent value: 1
Column 'ST':
  Unique values: 8
  Most frequent value: Oregon
  Count of most frequent value: 320
Column 'GENDER':
  Unique values: 5
  Most frequent value: F
  Count of most frequent value: 457
Column 'Education':
  Unique values: 6
  Most frequent value: Bachelor
  Count of most frequent value: 324
Column 'Number of Open Complaints':
  Unique values: 6
  Most frequent value: 1/0/00
  Count of most frequent value: 830
Column 'Policy Type':
  Unique values: 3
  Most frequent value: Personal Auto
  Count of most frequent value: 780
Column 'Vehicle Class':
  Unique values: 6
  Most frequent value: Four-Door Car
  Count of most frequent value: 576


The dataset includes customer and policy information. Here are some key findings:

There are 1071 unique customers, with "RB50392" as the most frequent identifier.
There are 8 unique states, with "Oregon" being the most common.
There are 5 unique genders, with "Female" (F) being the most frequent.
"Bachelor" is the most common education level among customers.
The "Number of Open Complaints" column may have data issues.
"Personal Auto" is the most common policy type.
"Four-Door Car" is the most frequent vehicle class.


In [22]:
total_claim_stats = df['total claim amount'].describe()


threshold_claim_amount = df['total claim amount'].quantile(0.75)

high_claim_customers = df[df['total claim amount'] > threshold_claim_amount]


summary_high_claim = high_claim_customers['total claim amount'].describe()


print("Summary Statistics for High Policy Claim Amount Data:")
print(summary_high_claim)

Summary Statistics for High Policy Claim Amount Data:
count     264.000000
mean      782.228263
std       292.751640
min       537.600000
25%       606.521741
50%       679.597985
75%       851.400000
max      2893.239678
Name: total claim amount, dtype: float64


In [23]:
high_claim_customers

customer          st gender education  customer lifetime value   income  \
1     QZ44356     Arizona      F  Bachelor                697953.59      0.0   
2     AI49188      Nevada      F  Bachelor               1288743.17  48767.0   
17    OE15005        Cali    NaN   College                394524.16  28855.0   
23    TZ98966      Nevada    NaN  Bachelor                245019.10      0.0   
26    US89481  California    NaN  Bachelor                394637.21      0.0   
...       ...         ...    ...       ...                      ...      ...   
1059  YG44474      Oregon      M   College               1401472.13  54193.0   
1061  RY92647        Cali      F  Bachelor               1050677.17      0.0   
1068  GS98873     Arizona      F  Bachelor                323912.47  16061.0   
1069  CW49887  California      F    Master                462680.11  79487.0   
1070  MY31220  California      F   College                899704.02  54230.0   

      monthly premium auto number of open complaints     policy type  \
1                     94.0                    1/0/00   Personal Auto   
2                    108.0                    1/0/00   Personal Auto   
17                   101.0                    1/0/00   Personal Auto   
23                    73.0                    1/3/00  Corporate Auto   
26                   111.0                    1/0/00   Personal Auto   
...                    ...                       ...             ...   
1059                 117.0                    1/0/00  Corporate Auto   
1061                  92.0                    1/0/00   Personal Auto   
1068                  88.0                    1/0/00   Personal Auto   
1069                 114.0                    1/0/00    Special Auto   
1070                 112.0                    1/0/00   Personal Auto   

      vehicle class  total claim amount  
1     Four-Door Car         1131.464935  
2      Two-Door Car          566.472247  
17              SUV          647.442031  
23    Four-Door Car          554.376763  
26    Four-Door Car          799.200000  
...             ...                 ...  
1059            SUV          720.752945  
1061  Four-Door Car          546.524896  
1068  Four-Door Car          633.600000  
1069            SUV          547.200000  
1070   Two-Door Car          537.600000  

[264 rows x 11 columns]

In [24]:

policy_counts = df['policy type'].value_counts()


policy_counts_df = pd.DataFrame(policy_counts, columns=['Total Policies Sold'])

policy_type_with_highest_sales = policy_counts.idxmax()

print("Total Number of Policies Sold for Each Policy Type:")
print(policy_counts_df)

print("Policy Type with the Highest Number of Policies Sold:", policy_type_with_highest_sales)


Total Number of Policies Sold for Each Policy Type:
Empty DataFrame
Columns: [Total Policies Sold]
Index: []
Policy Type with the Highest Number of Policies Sold: Personal Auto


In [25]:
personal_auto_df = df.loc[df['policy type'] == 'Personal Auto']
corporate_auto_df = df.loc[df['policy type'] == 'Corporate Auto']

average_income_personal_auto = personal_auto_df['income'].mean()
average_income_corporate_auto = corporate_auto_df['income'].mean()

print("Average Income for Personal Auto Policies:", average_income_personal_auto)
print("Average Income for Corporate Auto Policies:", average_income_corporate_auto)

Average Income for Personal Auto Policies: 38180.69871794872
Average Income for Corporate Auto Policies: 41390.31196581197
